odoo/addons/base/models/ir_qweb.py

```diff
@@ -639,72 +707,67 @@ def not_found_template(self, values, gen0=None):
+
             compile_context['_text_concat'] = []
             self._append_text("", compile_context) # To ensure the template function is a generator and doesn't become a regular function
-            code_lines = (
-                [f'def {def_name}(self, values, gen0, log):']
-                + self._compile_node(element, compile_context, 1)
-                + self._flush_text(compile_context, 1, rstrip=True)
-            )
-            for lines in compile_context['_all_def']:
-                code_lines.extend(lines)
+            compile_context['template_functions'][f'{def_name}_content'] = (
+                [f"def {def_name}_content(self, values):"]
+                + self._compile_node(element, compile_context, 2)
+                + self._flush_text(compile_context, 2, rstrip=True))
+
+            compile_context['template_functions'][def_name] = [indent_code(f"""
+                def {def_name}(self, values):
+                    try:
+                        if '__qweb_loaded_values' not in values:
+                            values['__qweb_loaded_values'] = {{}}
+                            values['__qweb_root_values'] = values.copy()
+                            values['xmlid'] = {options['ref_name']!r}
+                            values['viewid'] = {options['ref']!r}
+                        values['__qweb_loaded_values'].update(template_functions)
+
+                        yield from {def_name}_content(self, values)
+                    except QWebException:
+                        raise
+                    except Exception as e:
+                        raise QWebException("Error while render the template",
+                            self, template, ref={compile_context['ref']!r}, code=code) from e
+                    """, 0)]
         except QWebException:
             raise
         except Exception as e:
             raise QWebException("Error when compiling xml template",
-                self, template, path=compile_context.get('last_path_node')) from e
+                self, template, ref=compile_context['ref'], path_xml=compile_context['_qweb_error_path_xml']) from e
+
+        code_lines = ['code = None']
+        code_lines.append(f'template = {(document if isinstance(template, etree._Element) else template)!r}')
+        code_lines.append('template_functions = {}')
+
+        for lines in compile_context['template_functions'].values():
+            code_lines.extend(lines)
+
         try:
             code = '\n'.join(code_lines)
         except Exception as e:
             code = '\n'.join(map(str, code_lines))
             raise QWebException("Error when compiling xml template",
-                self, template, code=code) from e
+                self, template, ref=compile_context['ref'], code=code) from e
 
-        # compile code and defined default values
+        for name in compile_context['template_functions']:
+            code += f'\ntemplate_functions[{name!r}] = {name}'
```

Comment:
If a directive return a number instead of a string this method can fail. Before there were more cases, now it's not useful anymore, this during low level dev.